givenergy_modbus/client/commands.py

```diff
@@ -201,6 +385,22 @@ def set_battery_soc_reserve(val: int) -> list[TransparentRequest]:
         raise ValueError(f"Minimum SOC / shallow charge ({val}) must be in [4-100]%")
     return [WriteHoldingRegisterRequest(RegisterMap.BATTERY_SOC_RESERVE, val)]
 
+def set_car_charge_boost(val: int) -> list[TransparentRequest]:
+    """Set the minimum level of charge to maintain."""
+    # TODO what are valid values? 4-100?
+    val = int(val)
+    if not 0 <= val <= 22000:
+        raise ValueError(f"Charge Boost power ({val}) must be in [0-22kw]%")
+    return [WriteHoldingRegisterRequest(RegisterMap.CAR_CHARGE_BOOST, val)]
```

Comment:
Problem I now have is that CAR_CHARGE_BOOST appears to be an ems register, not an inverter register. Does the car charger appear as an EMS device ?  I don't know how all these things plug together.  Do they all just appear as different devices on one big modbus, or does the client have to connect specifically to the car charger, rather than the inverter, to be able to issue a CAR_CHARGE_BOOST ?

I'm aware that the car charger can be configured either as a standalone unit, or can be under the control of the inverter. Is this only used when it's configured separately ?

Anyway, I don't like the fact that this command is just sitting here and will send the request on any client connection, whether it's to a car charger or something else entirely.  Okay, it will probably get ignored if that register doesn't exist on the targetted device, but something still seems very wrong.

It seems to me that the commands ought to have some idea of what sort of device is connected, in order to be able to issue valid commands. I guess another case in point is that there seems to be EMS_DISCHARGE_SLOTS and DISCHARGE_SLOTS. So the appropriate request surely depends on what you're connected to ???

Reply:
Looking at the installation guides, Gateway manual suggests chargepoint can be plugged into it. Does the Gateway act as an EMS in this case ?

The chargepoint manual says that, still to come, the chargepoint can be wired into a gen-3 hybrid inverter. Would that be acting as an EMS in that case ?  Or just making it available on its modbus ?

Reply:
At the very least, lookup_writable_register() will need to know the device model type. But rather than just telling it "look for this Inverter register", it would be better to go via the plant instance, to use the model detected (or configured). At that almost certainly means going via the client instance.

So simplest is just to move the set_* commands (back ?) into client.py as instance methods. I'm not sure why the design was such that commands returned a list of pdu requests which you then had to send to client.exec() to action. Seems to make more sense that  client.set_charging_limit(x) just goes ahead and sends the request. Which does then mean it has to become async, and caller has to await it.


On a related note, does it still make sense for plant to offer direct inverter acces (ie plant.inverter) ?  In the new world, we don't know if it's an inverter we're attached to, or somethng else.   Perhaps plant.device which will return an instance of Inverter, Gateway, or whatever else it's atached to.

(Does an EMS offer access to the devices attached to it, or do you have to speak to them separately ?)

Reply:
I'm on holiday currently so there's a lot to catch up on but the gateways have their own EMS in them (Energy Management System) 
Ontop of that is the new CEMS unit Giv are developing which takes multiple 'old inverters' and conne to to them via USB and then the CEMS unit has full control not the inverter - so you I terfave to the CEMS unit exclusively which then controls the inverters (if you talk to the inverters directly with a CEMS unit attached then bad things happen)

Reply:
Worth noting, that while the Gateway's have an EMS, there is a product called an EMS (soon to be replaced by the CEMS) which users will understand as a different product.
For an CEMS, as hoggy notes, it should be the only way to access the inverters at a modbus level. (although it will have a restriction that won't get battery cell level data...).
For a 1:1 Gateway, you can get the "normal" AIO registers at the normal addresses. Not sure about multiple inverters on a single Gateway. That I still need to get my head around (am currently testing someones)

Reply:
> so you I terfave to the CEMS unit exclusively which then controls the inverters (if you talk to the inverters directly with a CEMS unit attached then bad things happen)

I'm guessing it's not easy to talk to the inverters directly anyway, since the USB connection to the CEMS replaces the dongle ?

Reply:
Yes and no, on Gen 1 removal of the Dongle does somewhat put an end to direct local access (unless you gonin directly via RS485 which is something I've been playing with getting to work) 
On Gen2 onwards the dongle is built in HOWEVER they still have a USB port available - and you can infact put a Gen1 dongle into them and run them from that. 
Same for the AIO gateway you are not forced to use the built in dongle should you wish (although why you wouldn't is unclear other than some sort of failover if your LAN failed or something)